File: src/modules/packages.ts
import { exec } from '../utils/exec.js';
import { getDistroInfo } from '../utils/os.js';
import type { AnalysisResult, CleaningResult, PackageManager } from '../types/index.js';
import { logger } from '../utils/logger.js';
import { parseSize } from '../utils/fs.js';

export class PackagesModule {
  readonly id = 'packages';
  readonly name = 'Gerenciador de Pacotes';
  readonly description = 'Limpa cache e remove pacotes órfãos do sistema';

  private packageManager: PackageManager;

  constructor() {
    this.packageManager = getDistroInfo().packageManager;
  }

  isAvailable(): boolean {
    return this.packageManager !== 'unknown';
  }

  async analyze(): Promise<AnalysisResult> {
    switch (this.packageManager) {
      case 'apt':
        return this.analyzeApt();
      case 'pacman':
        return this.analyzePacman();
      case 'dnf':
        return this.analyzeDnf();
      default:
        return { module: this.id, items: [], totalSize: 0 };
    }
  }

  private async analyzeApt(): Promise<AnalysisResult> {
    const items: AnalysisResult['items'] = [];
    let totalSize = 0;

    try {
      const cacheResult = await exec('du', ['-sb', '/var/cache/apt/archives']);
      if (cacheResult.success) {
        const match = cacheResult.stdout.match(/^(\d+)/);
        if (match) {
          const size = parseInt(match[1], 10);
          items.push({
            path: '/var/cache/apt/archives',
            size,
            type: 'apt-cache',
            description: 'Cache do APT (/var/cache/apt/archives)',
          });
          totalSize += size;
        }
      }
    } catch {
      items.push({
        path: '/var/cache/apt/archives',
        size: 0,
        type: 'apt-cache',
        description: 'Cache do APT (sem permissão para calcular)',
      });
    }

    try {
      const orphansResult = await exec('apt', ['autoremove', '--dry-run']);
      if (orphansResult.success) {
        const match = orphansResult.stdout.match(/(\d+)\s+package/i);
        if (match) {
          const count = parseInt(match[1], 10);
          items.push({
            path: 'apt-orphans',
            size: count * 50 * 1024 * 1024,
            type: 'apt-orphans',
            description: `Pacotes órfãos APT (${count})`,
          });
          totalSize += count * 50 * 1024 * 1024;
        }
      }
    } catch {
      // Dry-run pode falhar se não houver órfãos
    }

    return { module: this.id, items, totalSize };
  }

  private async analyzePacman(): Promise<AnalysisResult> {
    const items: AnalysisResult['items'] = [];
    let totalSize = 0;

    const cacheResult = await exec('du', ['-sb', '/var/cache/pacman/pkg']);
    if (cacheResult.success) {
      const match = cacheResult.stdout.match(/^(\d+)/);
      if (match) {
        const size = parseInt(match[1], 10);
        items.push({
          path: '/var/cache/pacman/pkg',
          size,
          type: 'pacman-cache',
          description: 'Cache do Pacman (/var/cache/pacman/pkg)',
        });
        totalSize += size;
      }
    }

    const orphansResult = await exec('paccache', ['-dk0']);
    if (orphansResult.success) {
      const match = orphansResult.stdout.match(/([\d.]+)\s*(?:KiB|MiB|GiB)/i);
      if (match) {
        const size = parseSize(match[1] + ' KiB');
        items.push({
          path: 'pacman-orphans',
          size,
          type: 'pacman-orphans',
          description: 'Versões antigas no cache (paccache -dk0)',
        });
        totalSize += size;
      }
    }

    return { module: this.id, items, totalSize };
  }

  private async analyzeDnf(): Promise<AnalysisResult> {
    const items: AnalysisResult['items'] = [];
    let totalSize = 0;

    const cacheResult = await exec('du', ['-sb', '/var/cache/dnf']);
    if (cacheResult.success) {
      const match = cacheResult.stdout.match(/^(\d+)/);
      if (match) {
        const size = parseInt(match[1], 10);
        items.push({
          path: '/var/cache/dnf',
          size,
          type: 'dnf-cache',
          description: 'Cache do DNF (/var/cache/dnf)',
        });
        totalSize += size;
      }
    }

    return { module: this.id, items, totalSize };
  }

  async clean(dryRun: boolean = false): Promise<CleaningResult> {
    const analysis = await this.analyze();
    const result: CleaningResult = {
      module: this.id,
      success: true,
      spaceFreed: 0,
      itemsRemoved: 0,
      errors: [],
    };

    if (dryRun) {
      logger.info(`[DRY-RUN] ${this.name}: limparía ${logger.formatBytes(analysis.totalSize)}`);
      result.spaceFreed = analysis.totalSize;
      return result;
    }

    switch (this.packageManager) {
      case 'apt':
        return this.cleanApt(analysis);
      case 'pacman':
        return this.cleanPacman(analysis);
      case 'dnf':
        return this.cleanDnf(analysis);
      default:
        result.errors.push('Gerenciador de pacotes não suportado');
        return result;
    }
  }

  private async cleanApt(analysis: AnalysisResult): Promise<CleaningResult> {
    const beforeSize = analysis.totalSize;
    const result: CleaningResult = {
      module: this.id,
      success: true,
      spaceFreed: 0,
      itemsRemoved: 0,
      errors: [],
    };

    const cleanResult = await exec('apt', ['clean'], { sudo: true });
    if (cleanResult.success) {
      logger.item(`APT: Cache limpo`);
      result.itemsRemoved++;
    } else {
      result.errors.push('Falha ao limpar cache APT (verifique se tem privilégios sudo)');
    }

    const autoremoveResult = await exec('apt', ['autoremove', '-y'], { sudo: true });
    if (autoremoveResult.success) {
      logger.item(`APT: Pacotes órfãos removidos`);
      result.itemsRemoved++;
    }

    const afterAnalysis = await this.analyzeApt();
    result.spaceFreed = Math.max(0, beforeSize - afterAnalysis.totalSize);

    return result;
  }

  private async cleanPacman(analysis: AnalysisResult): Promise<CleaningResult> {
    const beforeSize = analysis.totalSize;
    const result: CleaningResult = {
      module: this.id,
      success: true,
      spaceFreed: 0,
      itemsRemoved: 0,
      errors: [],
    };

    const cleanResult = await exec('pacman', ['-Sc', '--noconfirm'], { sudo: true });
    if (cleanResult.success) {
      logger.item(`Pacman: Cache limpo (mantém última versão)`);
      result.itemsRemoved++;
    } else {
      result.errors.push('Falha ao limpar cache Pacman (verifique se tem privilégios sudo)');
    }

    try {
      const orphansResult = await exec('pacman', ['-Qtdq']);
      if (orphansResult.success && orphansResult.stdout.trim()) {
        const orphanPackages = orphansResult.stdout
          .trim()
          .split('\n')
          .filter(p => p);
        for (const pkg of orphanPackages) {
          const removeResult = await exec('pacman', ['-Rns', pkg, '--noconfirm'], { sudo: true });
          if (removeResult.success) {
            result.itemsRemoved++;
          }
        }
        if (orphanPackages.length > 0) {
          logger.item(`Pacman: ${orphanPackages.length} pacotes órfãos removidos`);
        }
      }
    } catch {
      logger.debug('Nenhum pacote órfão encontrado');
    }

    const afterAnalysis = await this.analyzePacman();
    result.spaceFreed = Math.max(0, beforeSize - afterAnalysis.totalSize);

    return result;
  }

  private async cleanDnf(analysis: AnalysisResult): Promise<CleaningResult> {
    const beforeSize = analysis.totalSize;
    const result: CleaningResult = {
      module: this.id,
      success: true,
      spaceFreed: 0,
      itemsRemoved: 0,
      errors: [],
    };

    const cleanResult = await exec('dnf', ['clean', 'all'], { sudo: true });
    if (cleanResult.success) {
      logger.item(`DNF: Cache limpo`);
      result.itemsRemoved++;
    } else {
      result.errors.push('Falha ao limpar cache DNF (verifique se tem privilégios sudo)');
    }

    const autoremoveResult = await exec('dnf', ['autoremove', '-y'], { sudo: true });
    if (autoremoveResult.success) {
      logger.item(`DNF: Pacotes órfãos removidos`);
      result.itemsRemoved++;
    }

    const afterAnalysis = await this.analyzeDnf();
    result.spaceFreed = Math.max(0, beforeSize - afterAnalysis.totalSize);

    return result;
  }
}

export const packagesModule = new PackagesModule();

File: internal/modules/packages.go
package modules

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/joaomjbraga/piunter/internal/utils"
	"github.com/joaomjbraga/piunter/pkg/types"
)

type PackagesModule struct {
	BaseModule
	packageManager string
}

func NewPackagesModule() *PackagesModule {
	return &PackagesModule{
		BaseModule: BaseModule{
			id:          "packages",
			name:        "Pacotes Órfãos",
			description: "Remove pacotes órfãos do sistema",
		},
		packageManager: detectPackageManager(),
	}
}

func detectPackageManager() string {
	if utils.IsCommandAvailable("apt") {
		return "apt"
	}
	if utils.IsCommandAvailable("pacman") {
		return "pacman"
	}
	if utils.IsCommandAvailable("dnf") {
		return "dnf"
	}
	return ""
}

func (m *PackagesModule) IsAvailable() bool {
	return m.packageManager != "" && utils.IsCommandAvailable(m.packageManager)
}

func getAptSizes(items []types.CleanableItem) {
	executor := utils.GetExecutor()
	for i, item := range items {
		result := executor.Exec("dpkg-query", "-W", "-f=${Installed-Size}", item.Path)
		if result.Success {
			sizeKB, err := strconv.ParseInt(strings.TrimSpace(result.Stdout), 10, 64)
			if err == nil {
				items[i].Size = sizeKB * 1024
			}
		}
	}
}

func getPacmanSizes(items []types.CleanableItem) {
	executor := utils.GetExecutor()
	for i, item := range items {
		result := executor.Exec("pacman", "-Si", item.Path)
		if !result.Success {
			continue
		}
		for _, line := range strings.Split(result.Stdout, "\n") {
			if strings.HasPrefix(line, "Installed Size") {
				parts := strings.SplitN(line, ":", 2)
				if len(parts) == 2 {
					items[i].Size = parseSize(strings.TrimSpace(parts[1]))
				}
				break
			}
		}
	}
}

func getDnfSizes(items []types.CleanableItem) {
	executor := utils.GetExecutor()
	for i, item := range items {
		result := executor.Exec("dnf", "info", item.Path)
		if !result.Success {
			continue
		}
		for _, line := range strings.Split(result.Stdout, "\n") {
			line = strings.TrimSpace(line)
			if strings.HasPrefix(line, "Size") {
				parts := strings.SplitN(line, ":", 2)
				if len(parts) == 2 {
					items[i].Size = parseSize(strings.TrimSpace(parts[1]))
				}
				break
			}
		}
	}
}

func (m *PackagesModule) Analyze(threshold int) (*types.AnalysisResult, error) {
	result := &types.AnalysisResult{
		Module:    m.id,
		Items:    []types.CleanableItem{},
		TotalSize: 0,
	}

	var cmd string
	var args []string

	switch m.packageManager {
	case "apt":
		cmd = "apt-get"
		args = []string{"--just-print", "autoremove"}
	case "pacman":
		cmd = "pacman"
		args = []string{"-Qttd"}
	case "dnf":
		cmd = "dnf"
		args = []string{"autoremove", "--assumeno"}
	}

	executor := utils.GetExecutor()
	execResult := executor.Exec(cmd, args...)
	if !execResult.Success {
		if execResult.Code != 1 {
			return result, utils.NewAnalysisError(m.id, fmt.Sprintf("falha ao listar pacotes órfãos com %s", m.packageManager), fmt.Errorf("%s", execResult.Stderr))
		}
		stderr := strings.ToLower(execResult.Stderr)
		if strings.Contains(stderr, "e:") || strings.Contains(stderr, "error") {
			return result, utils.NewAnalysisError(m.id, fmt.Sprintf("falha ao listar pacotes órfãos com %s", m.packageManager), fmt.Errorf("%s", execResult.Stderr))
		}
	}

	lines := strings.Split(execResult.Stdout, "\n")
	orphanCount := 0

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		switch m.packageManager {
		case "apt":
			if !strings.HasPrefix(line, "Remv") {
				continue
			}
			parts := strings.Fields(line)
			if len(parts) >= 2 {
				orphanCount++
				result.Items = append(result.Items, types.CleanableItem{
					Path:        parts[1],
					Size:        0,
					Type:        "package",
					Description: "Pacote órfão: " + parts[1],
				})
			}
		case "pacman":
			parts := strings.Fields(line)
			if len(parts) >= 1 {
				orphanCount++
				result.Items = append(result.Items, types.CleanableItem{
					Path:        parts[0],
					Size:        0,
					Type:        "package",
					Description: "Pacote órfão: " + parts[0],
				})
			}
		case "dnf":
			if len(line) > 0 && (line[0] == ' ' || line[0] == '\t') {
				fields := strings.Fields(line)
				if len(fields) >= 1 && fields[0] != "==" && fields[0] != "--" {
					orphanCount++
					result.Items = append(result.Items, types.CleanableItem{
						Path:        fields[0],
						Size:        0,
						Type:        "package",
						Description: "Pacote órfão: " + fields[0],
					})
				}
			}
		}
	}

	if orphanCount > 0 {
		switch m.packageManager {
		case "apt":
			getAptSizes(result.Items)
		case "pacman":
			getPacmanSizes(result.Items)
		case "dnf":
			getDnfSizes(result.Items)
		}

		var total int64
		for _, item := range result.Items {
			total += item.Size
		}
		if total == 0 {
			total = int64(orphanCount) * 10 * utils.MB
		}
		result.TotalSize = total
	}

	return result, nil
}

func (m *PackagesModule) Clean(dryRun bool) (*types.CleaningResult, error) {
	analysis, err := m.Analyze(0)
	if err != nil {
		return &types.CleaningResult{
			Module:  m.id,
			Success: false,
			Errors:  []string{err.Error()},
		}, err
	}

	result := &types.CleaningResult{
		Module:       m.id,
		Success:      true,
		SpaceFreed:   0,
		ItemsRemoved: 0,
		Errors:       []string{},
	}

	if analysis.TotalSize == 0 {
		return result, nil
	}

	if dryRun {
		result.SpaceFreed = analysis.TotalSize
		result.ItemsRemoved = len(analysis.Items)
		utils.Info(fmt.Sprintf("[DRY-RUN] Limparia %d pacotes órfãos", len(analysis.Items)))
		return result, nil
	}

	executor := utils.GetExecutor()
	var execResult types.CommandResult

	switch m.packageManager {
	case "apt":
		execResult = executor.Exec("sudo", "apt", "autoremove", "-y")
	case "pacman":
		var pkgNames []string
		for _, item := range analysis.Items {
			pkgNames = append(pkgNames, item.Path)
		}
		if len(pkgNames) == 0 {
			return result, nil
		}
		args := append([]string{"pacman", "-Rns", "--noconfirm"}, pkgNames...)
		execResult = executor.Exec("sudo", args...)
	case "dnf":
		execResult = executor.Exec("sudo", "dnf", "autoremove", "-y")
	}

	if execResult.Success {
		result.SpaceFreed = analysis.TotalSize
		result.ItemsRemoved = len(analysis.Items)
		utils.Item(m.Name(), fmt.Sprintf("%d pacotes removidos", len(analysis.Items)))
	} else {
		result.Success = false
		result.Errors = append(result.Errors, fmt.Sprintf("Falha ao limpar pacotes: %s", execResult.Stderr))
	}

	return result, nil
}